File: 152.maximum-product-subarray.py
# @leet start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        mini = nums[0]
        maxi = nums[0]
        answer = nums[0]
        for i in range(1, len(nums)):
            temp = min(nums[i], nums[i] * mini, nums[i] * maxi)
            maxi = max(nums[i], nums[i] * mini, nums[i] * maxi)
            mini = temp
            answer = max(answer, maxi)
        return answer


# @leet end

File: 152.maximum-product-subarray.java
// @leet start
class Solution {
    public int maxProduct(int[] nums) {
		int mini = nums[0];
		int maxi = nums[0];
		int answer  = nums[0];
		for(int i = 1; i< nums.length; i++){
			int temp = Math.min(nums[i],  Math.min(nums[i] * mini, nums[i] * maxi));
			maxi = Math.max(nums[i],  Math.max(nums[i] * mini, nums[i] * maxi));
			mini = temp;
			answer = Math.max(answer, maxi);
		}
		return answer;
    }
}
// @leet end
